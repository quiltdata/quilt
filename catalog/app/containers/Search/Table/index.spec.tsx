import * as React from 'react'
import { render } from '@testing-library/react'

import type { SearchHitPackage } from '../model'

import TableView from './index'

jest.mock('components/Layout', () => ({
  useSetFullWidth: jest.fn(),
}))

jest.mock('../model', () => ({
  use: () => ({
    state: {
      resultType: 'p',
      view: 't',
    },
  }),
  ResultType: {
    QuiltPackage: 'p',
    S3Object: 'o',
  },
  View: {
    Table: 't',
    List: 'l',
  },
}))

const useResults = jest.fn()
jest.mock('./useResults', () => ({
  useResults: () => useResults(),
}))

jest.mock('../NoResults', () => ({
  Skeleton: () => <div>Loadingâ€¦</div>,
  Error: ({ children, kind }: { children?: React.ReactNode; kind?: string }) => {
    switch (kind) {
      case 'syntax':
        return (
          <section>
            <h1>Syntax error</h1>
            <details>{children}</details>
          </section>
        )
      case 'timeout':
        return (
          <section>
            <h1>Timeout error</h1>
            <details>{children}</details>
          </section>
        )
      default:
        return (
          <section>
            <h1>Unexpected error</h1>
            <details>{children}</details>
          </section>
        )
    }
  },
}))

jest.mock('./Table', () => ({ hits }: { hits: SearchHitPackage[] }) => (
  <table>
    <tbody>
      {hits.map((hit) => (
        <tr key={hit.id}>
          <td>{hit.name}</td>
        </tr>
      ))}
    </tbody>
  </table>
))

const TablePage = () => (
  <TableView emptySlot={<div>No results</div>} onRefine={jest.fn()} />
)

describe('containers/Search/Table/index', () => {
  describe('when no results', () => {
    beforeEach(() => {
      jest.clearAllMocks()
    })

    it('renders null for idle state', () => {
      useResults.mockReturnValue([{ _tag: 'idle' }])

      const { container } = render(<TablePage />)
      expect(container).toMatchSnapshot()
    })

    it('renders skeleton for in-progress state', () => {
      useResults.mockReturnValue([{ _tag: 'in-progress' }])

      const { container } = render(<TablePage />)
      expect(container).toMatchSnapshot()
    })

    it('renders error for general fail state', () => {
      useResults.mockReturnValue([
        {
          _tag: 'fail',
          error: {
            _tag: 'general',
            error: new Error('Something went wrong'),
          },
        },
      ])

      const { container } = render(<TablePage />)
      expect(container).toMatchSnapshot()
    })

    it('renders error for page fail state', () => {
      useResults.mockReturnValue([
        {
          _tag: 'fail',
          error: {
            _tag: 'page',
            error: new Error('Page loading failed'),
          },
        },
      ])

      const { container } = render(<TablePage />)
      expect(container).toMatchSnapshot()
    })

    it('renders error for InputError data fail state', () => {
      useResults.mockReturnValue([
        {
          _tag: 'fail',
          error: {
            _tag: 'data',
            error: {
              __typename: 'InputError',
              message: 'We can not handle this input',
            },
          },
        },
      ])

      const { container } = render(<TablePage />)
      expect(container).toMatchSnapshot()
    })

    it('renders error for QuerySyntaxError data fail state', () => {
      useResults.mockReturnValue([
        {
          _tag: 'fail',
          error: {
            _tag: 'data',
            error: {
              __typename: 'InputError',
              name: 'QuerySyntaxError',
              path: 'search.query',
              message: 'Syntax error in query',
            },
          },
        },
      ])

      const { container } = render(<TablePage />)
      expect(container).toMatchSnapshot()
    })

    it('renders timeout error for OperationError data fail state', () => {
      useResults.mockReturnValue([
        {
          _tag: 'fail',
          error: {
            _tag: 'data',
            error: {
              __typename: 'OperationError',
              name: 'Timeout',
              message: 'Request timeout',
            },
          },
        },
      ])

      const { container } = render(<TablePage />)
      expect(container).toMatchSnapshot()
    })

    it('renders operation error for other OperationError data fail state', () => {
      useResults.mockReturnValue([
        {
          _tag: 'fail',
          error: {
            _tag: 'data',
            error: {
              __typename: 'OperationError',
              message: 'Internal server error',
            },
          },
        },
      ])

      const { container } = render(<TablePage />)
      expect(container).toMatchSnapshot()
    })

    it('renders empty slot for empty state', () => {
      useResults.mockReturnValue([{ _tag: 'empty' }])

      const { container } = render(<TablePage />)
      expect(container).toMatchSnapshot()
    })
  })

  describe('when has results', () => {
    it('renders table with rows', () => {
      useResults.mockReturnValue([
        {
          _tag: 'ok',
          hits: [
            { id: '1', name: 'package-1' },
            { id: '2', name: 'package-2' },
          ],
          cursor: null,
          determinate: true,
        },
      ])

      const { container } = render(<TablePage />)
      expect(container).toMatchSnapshot()
    })
  })
})
