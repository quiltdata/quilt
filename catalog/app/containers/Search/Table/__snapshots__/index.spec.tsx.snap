// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Search/Table/index when has results renders table with rows 1`] = `
<div>
  <div>
    <table>
      <tbody>
        <tr>
          <td>
            package-1
          </td>
        </tr>
        <tr>
          <td>
            package-2
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`containers/Search/Table/index when no results renders empty slot for empty state 1`] = `
<div>
  <div>
    No results
  </div>
</div>
`;

exports[`containers/Search/Table/index when no results renders error for InputError data fail state 1`] = `
<div>
  <section>
    <h1>
      Unexpected error
    </h1>
    <details>
      Invalid input at 
      <code />
      : 
      <pre
        style="white-space: pre-wrap;"
      >
        We can not handle this input
      </pre>
    </details>
  </section>
</div>
`;

exports[`containers/Search/Table/index when no results renders error for QuerySyntaxError data fail state 1`] = `
<div>
  <section>
    <h1>
      Syntax error
    </h1>
    <details>
      Invalid input at 
      <code>
        search.query
      </code>
      : 
      QuerySyntaxError
      <pre
        style="white-space: pre-wrap;"
      >
        Syntax error in query
      </pre>
    </details>
  </section>
</div>
`;

exports[`containers/Search/Table/index when no results renders error for general fail state 1`] = `
<div>
  <section>
    <h1>
      Unexpected error
    </h1>
    <details>
      Something went wrong
    </details>
  </section>
</div>
`;

exports[`containers/Search/Table/index when no results renders error for page fail state 1`] = `
<div>
  <section>
    <h1>
      Unexpected error
    </h1>
    <details>
      Page loading failed
    </details>
  </section>
</div>
`;

exports[`containers/Search/Table/index when no results renders null for idle state 1`] = `<div />`;

exports[`containers/Search/Table/index when no results renders operation error for other OperationError data fail state 1`] = `
<div>
  <section>
    <h1>
      Unexpected error
    </h1>
    <details>
      Operation error: 
      Internal server error
    </details>
  </section>
</div>
`;

exports[`containers/Search/Table/index when no results renders skeleton for in-progress state 1`] = `
<div>
  <div>
    Loadingâ€¦
  </div>
</div>
`;

exports[`containers/Search/Table/index when no results renders timeout error for OperationError data fail state 1`] = `
<div>
  <section>
    <h1>
      Timeout error
    </h1>
    <details />
  </section>
</div>
`;
